enhancements/cluster-logging/forward_to_cloudwatch.md

```diff
@@ -0,0 +1,198 @@
+- `groupLabel`: (string, *nice-to-have*) Kubernetes label to use as group name.
+  If the label is absent, use value of `group`.
+- `fetchIntervalSeconds`: (number, *nice-to-have*) interval to fetch logs in seconds.
+
+Existing fields:
+
+- `url`: Not used in production. Can be set for testing purposes.
+- `secret`: AWS credentials, keys `aws_access_key_id` and `aws_secret_access_key`.
+
+*Priority*: The `group`, `groupLabel` and `fetchIntervalSeconds` fields are lower priority.
+
+### User Stories
+
+#### I want to forward logs to Cloudwatch instead of a local store
+
+```
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+spec:
+  outputs:
+   - name: CloudwatchOut
+     type: cloudwatch
+     cloudwatch:
+       region: myregion
+     secret:
+        name: mysecret
+  pipelines:
+  - inputRefs: [application, infrastructure, audit]
+  - outputRefs: [CloudwatchOut]
+```
+
+[Cloudwatch][concepts] defines Log Groups and Log Streams.
+The CRD above creates 3 Log Groups named "application", "infrastructure" and "audit".
+A new Log Stream is created for each container producing logs.
+
+**Note**: The installer UI (Addon or OLM) can get AWS credentials from a `cloudcredential.openshift.io/v1`.
+The user only has to provide a `region` to enable cloudwatch forwarding for a cluster.
+Details are out of scope for this proposal.
+
+#### I want to define my own log groups
+
+There are two ways to define custom log groups:
+- Use a Kubernetes label with the group name.
+- Define the group name in the `ClusterLogForwarder.Output`
+
+1. Using a kubernetes label named `LogGroup`
+
+```
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+spec:
+  outputs:
+   - name: CloudwatchOut
+     type: cloudwatch
+     cloudwatch:
+       region: myregion
+	   groupLabel: LogGroup
+     secret:
+        name: mysecret
+  pipelines:
+  - inputRefs: [application, infrastructure, audit]
+  - outputRefs: [CloudwatchOut]
+```
+
+2. Defining a group in the output spec and use input selectors:
+
+```
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+  inputs:
+  - name: FooIn
+    application:
+	  namespaces: [foo]
+  - name: BarIn
+    application:
+	  namespaces: [bar]
+  outputs:
+   - name: FooOut
+     type: cloudwatch
+     cloudwatch:
+	   group: foo
+   - name: BarOut
+     type: cloudwatch
+     cloudwatch:
+	   group: bar
+  pipelines:
+  - inputRefs: [FooIn]
+    outputRefs: [FooOut]
+  - inputRefs: [BarIn]
+    outputRefs: [BarOut]
+```
+
+#### I want to use EKS authentication with Cloudwatch
+
+TODO: is this a requirement? If so define an appropriate set of `secret` keys.
+
+### Implementation Details
+
+Using the [fluentd cloudwatch plugin][plugin].
+Most of the CRD fields have an obvious corresponding fluentd parameter.
+
+Set `auto_create_stream true` to create streams and groups on the fly.
+
+`log-stream-name` is set to the `containerID` of the log record.
```

Comment:
Is the `containerID` the UUID of the container or the more human-readable name?

Reply:
From a usability point of view, how easy will it be for log consumers to identify which of these streams map to which applications?  Clarity on what `containerID` actually is in terms of mapping to metadata in-cluster will help clarify this.  If it's the UUID of the container it's a non-starter.  The name must map to something that isn't ephemeral.

Reply:
logstream per container would be easier to digest if the name was something similar to the fluent tag or the info really we can get from it:
* namespace
* podname
* container ID

Given the containerID is a hash that is more or less meaningless to users I would likely drop this from the stream unless there is a need to uniquely distinguish multi-container pod logs and there is no other mechanism

Reply:
The log entries contain all the metadata that the application might need (container name, uuid, namespace, k8s labels and more) The log stream name only needs to be a unique identifier that lives as long as the source of logs. I'm proposing a log-stream per container, so container UUID fits the bill.

Reply:
Before locking this down let's be sure from a UX point of view I can get to the logs with just the namespace or container name.  If I first have to know the stream in which to search and the stream is the UUID this is not going to work.